File: app/src/main/java/org/liebald/android/cula/data/CulaRepository.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.liebald.android.cula.data;

import android.annotation.SuppressLint;
import android.arch.lifecycle.LiveData;
import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import com.firebase.jobdispatcher.FirebaseJobDispatcher;
import com.firebase.jobdispatcher.GooglePlayDriver;
import com.firebase.jobdispatcher.Job;
import com.firebase.jobdispatcher.Lifetime;
import com.firebase.jobdispatcher.RetryStrategy;
import com.firebase.jobdispatcher.Trigger;

import org.liebald.android.cula.data.database.CulaDatabase;
import org.liebald.android.cula.data.database.Dao.LanguageDao;
import org.liebald.android.cula.data.database.Dao.LessonDao;
import org.liebald.android.cula.data.database.Dao.LibraryDao;
import org.liebald.android.cula.data.database.Dao.QuoteDao;
import org.liebald.android.cula.data.database.Entities.LanguageEntry;
import org.liebald.android.cula.data.database.Entities.LessonEntry;
import org.liebald.android.cula.data.database.Entities.LessonMappingEntry;
import org.liebald.android.cula.data.database.Entities.LibraryEntry;
import org.liebald.android.cula.data.database.Entities.MappingPOJO;
import org.liebald.android.cula.data.database.Entities.QuoteEntry;
import org.liebald.android.cula.services.UpdateQuoteJobService;
import org.liebald.android.cula.utilities.AppExecutors;

import java.util.List;

/**
 * Handles data operations in Cula.
 */
public class CulaRepository {

    /**
     * Tag for logging.
     */
    private static final String TAG = CulaRepository.class.getSimpleName();

    // For Singleton instantiation
    private static final Object LOCK = new Object();
    @SuppressLint("StaticFieldLeak")
    private static CulaRepository sInstance;
    private final LibraryDao mLibraryDao;
    private final LanguageDao mLanguageDao;
    private final QuoteDao mQuoteDao;
    private final LessonDao mLessonDao;
    private final AppExecutors mExecutors;
    private final SharedPreferences mSharedPreferences;
    private final Context mContext;


    private CulaRepository(CulaDatabase database, AppExecutors appExecutors, SharedPreferences sharedPreferences, Context context) {
        mLibraryDao = database.libraryDao();
        mExecutors = appExecutors;
        mLanguageDao = database.languageDao();
        mQuoteDao = database.quoteDao();
        mLessonDao = database.lessonDao();
        mSharedPreferences = sharedPreferences;
        mContext = context;

        //TODO: remove following testcode before publishing
        setDebugState();


        scheduleJobService();
    }

    /**
     * Singleton to make sure only one {@link CulaRepository} is used at a time.
     *
     * @param database          The {@link CulaDatabase} to access all {@link android.arch.persistence.room.Dao}s.
     * @param appExecutors      The {@link AppExecutors} used to execute all kind of queries of the main thread.
     * @param sharedPreferences The {@link SharedPreferences} used access the apps settings.
     * @return A new {@link CulaRepository} if none exists. If already an instance exists this is returned instead of creating a new one.
     */
    public synchronized static CulaRepository getInstance(
            CulaDatabase database, AppExecutors appExecutors, SharedPreferences sharedPreferences, Context context) {
        if (sInstance == null) {
            synchronized (LOCK) {
                sInstance = new CulaRepository(database, appExecutors, sharedPreferences, context);
                Log.d(TAG, "Made new repository");
            }
        }
        return sInstance;
    }


    /**
     * Schedules a Job service to regularly update the motivational quote.
     */
    private void scheduleJobService() {

        FirebaseJobDispatcher dispatcher = new FirebaseJobDispatcher(new GooglePlayDriver(mContext));

        // Update the database immediately
        Job myInstantJob = dispatcher.newJobBuilder()
                .setService(UpdateQuoteJobService.class)
                .setTag("updateQuoteJobServiceNow")
                .build();
        dispatcher.mustSchedule(myInstantJob);

        // and also update it each 12-13 hours as recurring job
        Job recurringJob = dispatcher.newJobBuilder()
                .setService(UpdateQuoteJobService.class)
                .setTag("updateQuoteJobServiceRecurrent")
                .setLifetime(Lifetime.FOREVER)
                .setRecurring(true)
                .setTrigger(Trigger.executionWindow(
                        12 * 60 * 60,
                        13 * 60 * 60))
                .setReplaceCurrent(true)
                .setRetryStrategy(RetryStrategy.DEFAULT_LINEAR)
                .build();
        dispatcher.mustSchedule(recurringJob);
    }

    /**
     * For debugging purposes prefill database with specified data.
     */
    private void setDebugState() {
        mExecutors.diskIO().execute(mLibraryDao::deleteAll);
        insertLanguageEntry(new LanguageEntry("German"));
        insertLanguageEntry(new LanguageEntry("Greek"));
        insertLibraryEntry(new LibraryEntry(1, "native1", "foreign", "German", 1.1));
        insertLibraryEntry(new LibraryEntry(2, "native2", "foreign2", "German", 2.2));
        insertLibraryEntry(new LibraryEntry(3, "native3", "foreign3", "German", 3.3));
        insertLibraryEntry(new LibraryEntry(4, "native4", "foreign4", "German", 4.4));
        insertLibraryEntry(new LibraryEntry(5, "native5", "foreign5", "German", 4.8));
        insertLibraryEntry(new LibraryEntry(6, "native6", "foreign6", "Greek", 2));
        insertLibraryEntry(new LibraryEntry(7, "native7", "foreign7", "Greek", 4));
        insertQuoteEntry(new QuoteEntry(1, "TestQuote of the Day with a rather medium long text", "Stefan"));
        insertLessonEntry(new LessonEntry(1, "Test Lesson 1", "This lesson is for testing purposes", "German"));
        insertLessonEntry(new LessonEntry(2, "Test Lesson 2", "This lesson is for testing purposes", "Greek"));
        insertLessonEntry(new LessonEntry(3, "Test Lesson 3", "This lesson is for testing purposes", "Greek"));
        insertLessonMappingEntry(new LessonMappingEntry(1, 1, 1));
        insertLessonMappingEntry(new LessonMappingEntry(2, 1, 2));
        insertLessonMappingEntry(new LessonMappingEntry(3, 2, 3));
        insertLessonMappingEntry(new LessonMappingEntry(4, 3, 4));
        insertLessonMappingEntry(new LessonMappingEntry(5, 3, 5));

        //print the current entries in the db to the log console.
        mExecutors.diskIO().execute(() ->
                Log.d(CulaRepository.class.getSimpleName(), "Database has now " + mLibraryDao.getLibrarySize() + " library entries")
        );
        mExecutors.diskIO().execute(() ->
                Log.d(CulaRepository.class.getSimpleName(), "Database has now " + mLanguageDao.getAmountOfLanguages() + " language entries")
        );
        mExecutors.diskIO().execute(() ->
                Log.d(CulaRepository.class.getSimpleName(), "Database has now " + mLessonDao.getAmountOfLessons() + " lesson entries")
        );
        mExecutors.diskIO().execute(() ->
                Log.d(CulaRepository.class.getSimpleName(), "Database has now " + mLessonDao.getAmountOfLessonsMappings() + " lesson mapping entries")
        );

    }



    /**
     * Get all {@link LibraryEntry}s.
     *
     * @return All {@link LibraryEntry}s.
     */
    public LiveData<List<LibraryEntry>> getAllLibraryEntries() {
        //todo: find out how to use the string resource here instead of the hard coded key.
        String language = mSharedPreferences.getString("languages", "123");
        Log.d(TAG, "Retrieving all entries for selected Language: " + language);
        return mLibraryDao.getAllEntries(language);
    }

    /**
     * Get all {@link LibraryEntry}s.
     *
     * @return The {@link LibraryEntry} with the given ID.
     */
    public LiveData<LibraryEntry> getLibraryEntry(int id) {
        return mLibraryDao.getEntryById(id);
    }

    /**
     * Adds the given {@link LibraryEntry}s to the Database.
     *
     * @param libraryEntry One or more {@link LibraryEntry}s to add to the Database
     */
    public void insertLibraryEntry(LibraryEntry... libraryEntry) {
        mExecutors.diskIO().execute(() -> mLibraryDao.insertEntry(libraryEntry));
        for (LibraryEntry lib :
                libraryEntry) {
            Log.d(TAG, "Added Entry to db:" + lib.toString());
        }
    }

    /**
     * Removes the given {@link LibraryEntry}s from the Database.
     *
     * @param libraryEntry The @{@link LibraryEntry}s to remove from the Database
     */
    public void deleteLibraryEntry(LibraryEntry libraryEntry) {
        mExecutors.diskIO().execute(() -> mLibraryDao.deleteEntry(libraryEntry));
    }

    /**
     * Get all {@link LanguageEntry}s.
     *
     * @return All {@link LanguageEntry}s.
     */
    public LiveData<List<LanguageEntry>> getAllLanguageEntries() {
        return mLanguageDao.getAllEntries();
    }

    /**
     * Removes the given {@link LanguageEntry}s from the Database.
     *
     * @param languageEntry The @{@link LanguageEntry}s to remove from the Database
     */
    public void deleteLanguageEntry(LanguageEntry languageEntry) {
        mExecutors.diskIO().execute(() -> mLanguageDao.deleteEntry(languageEntry));
    }

    /**
     * Adds the given {@link LanguageEntry}s to the Database. If the language already exists, nothing is done.
     *
     * @param languageEntries One or more {@link LanguageEntry}s to add to the Database
     */
    public void insertLanguageEntry(LanguageEntry... languageEntries) {
        mExecutors.diskIO().execute(() -> mLanguageDao.insertEntry(languageEntries));
    }


    /**
     * Adds the given {@link QuoteEntry}s to the Database.
     *
     * @param quoteEntries One or more {@link QuoteEntry}s to add to the Database
     */
    public void insertQuoteEntry(QuoteEntry... quoteEntries) {
        mExecutors.diskIO().execute(() -> mQuoteDao.insertEntry(quoteEntries));
    }

    /**
     * Load a new {@link QuoteEntry} and return it wrapped in {@link LiveData}.
     *
     * @return The {@link LiveData} wrapped {@link QuoteEntry}
     */
    public LiveData<QuoteEntry> getQuote() {
        return mQuoteDao.getLatestEntry();
    }


    /**
     * Get all {@link LessonEntry}s.
     *
     * @return All {@link LessonEntry}s.
     */
    public LiveData<List<LessonEntry>> getAllLessonEntries() {
        //todo: find out how to use the string resource here instead of the hard coded key.
        String language = mSharedPreferences.getString("languages", "123");
        Log.d(TAG, "Retrieving all entries for selected Language: " + language);
        return mLessonDao.getAllEntries(language);
    }

    /**
     * Adds the given {@link LessonEntry}s to the Database. If the {@link LessonEntry} already exists, nothing is done.
     *
     * @param lessonEntries One or more {@link LessonEntry}s to add to the Database
     */
    public void insertLessonEntry(LessonEntry... lessonEntries) {
        mExecutors.diskIO().execute(() -> mLessonDao.insertEntry(lessonEntries));
    }

    /**
     * Removes the given {@link LessonEntry}s from the Database.
     *
     * @param lessonEntry The @{@link LessonEntry}s to remove from the Database
     */
    public void deleteLessonEntry(LessonEntry lessonEntry) {
        mExecutors.diskIO().execute(() -> mLessonDao.deleteEntry(lessonEntry));
    }

    /**
     * Get the {@link LessonEntry} with the given ID.
     *
     * @return The {@link LessonEntry} with the given ID.
     */
    public LiveData<LessonEntry> getLessonEntry(int id) {
        return mLessonDao.getEntryById(id);
    }

    /**
     * Adds the given {@link LessonMappingEntry}s to the Database. If the {@link LessonMappingEntry} already exists, nothing is done.
     *
     * @param lessonMappingEntries One or more {@link LanguageEntry}s to add to the Database
     */
    public void insertLessonMappingEntry(LessonMappingEntry... lessonMappingEntries) {
        mExecutors.diskIO().execute(() -> mLessonDao.insertMappingEntry(lessonMappingEntries));
    }

    /**
     * Deletes a mapping between a {@link LessonEntry} and a {@link LibraryEntry}.
     *
     * @param lessonMappingEntry The {@link LessonMappingEntry} describing the mapping.
     */
    public void deleteLessonMappingEntry(LessonMappingEntry lessonMappingEntry) {
        mExecutors.diskIO().execute(() -> mLessonDao.deleteMappingEntry(
                lessonMappingEntry.getLessonEntryId(), lessonMappingEntry.getLibraryEntryId()));

    }

    /**
     * Gets the List of {@link MappingPOJO}s in the lesson database table with the given id
     * based on the current language.
     *
     * @param id The lesson id for which the {@link List} of {@link MappingPOJO}s
     *           should be retrieved.
     * @return {@link LiveData} with the {@link List} of @{@link MappingPOJO}s.
     */
    public LiveData<List<MappingPOJO>> getMappingEntries(int id) {
        //todo: find out how to use the string resource here instead of the hard coded key.
        String language = mSharedPreferences.getString("languages", "123");
        return mLessonDao.getLessonMappingById(id, language);
    }


}

File: app/src/main/java/com/sliebald/cula/data/CulaRepository.kt
package com.sliebald.cula.data

import android.util.Log
import androidx.lifecycle.LiveData
import com.sliebald.cula.BuildConfig
import com.sliebald.cula.data.database.CulaDatabase
import com.sliebald.cula.data.database.entities.*
import com.sliebald.cula.data.database.pojos.*
import com.sliebald.cula.utilities.AppExecutors
import java.util.*

/**
 * Handles data operations in Cula.
 */
class CulaRepository private constructor(database: CulaDatabase, appExecutors: AppExecutors) {

    /**
     * A callback when one or more [LessonEntry]s were added to the database.
     * Contains the Ids of the new entries
     */
    interface OnLessonEntryAddedListener {
        fun onLessonEntryAdded(ids: LongArray)
    }

    private val mLibraryDao = database.libraryDao()
    private val mLanguageDao = database.languageDao()
    private val mLessonDao = database.lessonDao()
    private val mStatisticsDao = database.statisticsDao()
    private val mExecutors = appExecutors

    //TODO: Check how to use singletons in kotlin using the object keyword
    companion object {
        /**
         * Tag for logging.
         */
        private val TAG = CulaRepository::class.simpleName

        // For Singleton instantiation
        private val LOCK = Any()
        private var init = false
        private lateinit var sInstance: CulaRepository

        /**
         * Singleton to make sure only one [CulaRepository] is used at a time.
         *
         * @param database     The [CulaDatabase] to access all
         * [androidx.room.Dao]s.
         * @param appExecutors The [AppExecutors] used to execute all kind of queries of
         * the main thread.
         * @return A new [CulaRepository] if none exists. If already an instance exists this is
         * returned instead of creating a new one.
         */
        @JvmStatic
        @Synchronized
        fun getInstance(database: CulaDatabase, appExecutors: AppExecutors): CulaRepository {
            if (!init) {
                synchronized(LOCK) {
                    if (!init) {
                        init = true
                        sInstance = CulaRepository(database, appExecutors)
                        Log.d(TAG, "Made new repository")
                    }
                }
            }
            return sInstance
        }
    }

    init {
        //Adding test data to database for testing
        if (BuildConfig.DEBUG) {
            setDebugState()
        }
    }

    /**
     * For debugging purposes pre-fill database with specified data.
     */
    private fun setDebugState() {
        //  mExecutors.diskIO().execute(mLibraryDao::deleteAll);
        insertLanguageEntry(LanguageEntry("German", true))
        insertLanguageEntry(LanguageEntry("Greek", false))
        insertLibraryEntry(LibraryEntry("Bread", "Brot", "German", 1.1, Date(), 1))
        insertLibraryEntry(LibraryEntry("Apple", "Apfel", "German", 2.2, Date(), 2))
        insertLibraryEntry(LibraryEntry("Banana", "Banane", "German", 3.3, Date(), 3))
        insertLibraryEntry(LibraryEntry("Pear", "Birne", "German", 0.5, Date(), 4))
        insertLibraryEntry(LibraryEntry("Blackberry", "Brombeere", "German", 4.8, Date(), 5))
        insertLibraryEntry(LibraryEntry("native6", "foreign6", "Greek", 2.0, Date(), 6))
        insertLibraryEntry(LibraryEntry("native7", "foreign7", "Greek", 4.0, Date(), 7))
        insertLibraryEntry(LibraryEntry("native8", "foreign8", "Greek", 4.0, Date(), 8))
        insertLibraryEntry(LibraryEntry("Strawberry", "Erdbeere", "German", 1.0, Date(), 9))
        insertLibraryEntry(LibraryEntry("Peanut", "Erdnuss", "German", 4.0, Date(), 10))
        insertLibraryEntry(LibraryEntry("Raspberry", "Himbeeere", "German", 4.7, Date(), 11))
        insertLibraryEntry(LibraryEntry("Cherry", "Kirsche", "German", 2.1, Date(), 12))
        insertLibraryEntry(LibraryEntry("Bedroom", "Schlafzimmer", "German", 3.3, Date(), 13))
        insertLibraryEntry(LibraryEntry("Living room", "Wohnzimmer", "German", 3.8, Date(), 14))
        insertLibraryEntry(LibraryEntry("Bath", "Bad", "German", 1.6, Date(), 15))
        insertLibraryEntry(LibraryEntry("bath2", "bad2", "German", 1.6, Date(), 16))
        val dummyListener = object : OnLessonEntryAddedListener {
            override fun onLessonEntryAdded(ids: LongArray) {}
        }
        insertLessonEntry(dummyListener, LessonEntry("Fruits", "This lesson contains all fruits", "German", 1))
        insertLessonEntry(dummyListener, LessonEntry("Rooms", "This lesson contains rooms", "German", 2))
        insertLessonEntry(dummyListener, LessonEntry("Test Lesson 3", "This lesson is for testing purposes", "Greek", 3))
        insertLessonEntry(dummyListener, LessonEntry("Everything", "This lesson contains all German words", "German", 4))
        insertLessonEntry(dummyListener, LessonEntry("everything", "This lesson contains all German words", "German", 5))
        insertLessonMappingEntry(LessonMappingEntry(1, 1, 1))
        insertLessonMappingEntry(LessonMappingEntry(1, 2, 2))
        insertLessonMappingEntry(LessonMappingEntry(1, 3, 3))
        insertLessonMappingEntry(LessonMappingEntry(1, 4, 4))
        insertLessonMappingEntry(LessonMappingEntry(1, 5, 5))
        insertLessonMappingEntry(LessonMappingEntry(1, 9, 6))
        insertLessonMappingEntry(LessonMappingEntry(1, 10, 7))
        insertLessonMappingEntry(LessonMappingEntry(1, 11, 8))
        insertLessonMappingEntry(LessonMappingEntry(1, 12, 9))
        insertLessonMappingEntry(LessonMappingEntry(2, 13, 10))
        insertLessonMappingEntry(LessonMappingEntry(2, 14, 11))
        insertLessonMappingEntry(LessonMappingEntry(2, 15, 12))
        insertLessonMappingEntry(LessonMappingEntry(4, 1, 13))
        insertLessonMappingEntry(LessonMappingEntry(4, 2, 14))
        insertLessonMappingEntry(LessonMappingEntry(4, 3, 15))
        insertLessonMappingEntry(LessonMappingEntry(4, 4, 16))
        insertLessonMappingEntry(LessonMappingEntry(4, 5, 17))
        insertLessonMappingEntry(LessonMappingEntry(4, 9, 18))
        insertLessonMappingEntry(LessonMappingEntry(4, 10, 19))
        insertLessonMappingEntry(LessonMappingEntry(4, 11, 20))
        insertLessonMappingEntry(LessonMappingEntry(4, 12, 21))
        insertLessonMappingEntry(LessonMappingEntry(4, 13, 22))
        insertLessonMappingEntry(LessonMappingEntry(4, 14, 23))
        insertLessonMappingEntry(LessonMappingEntry(4, 15, 24))
        insertLessonMappingEntry(LessonMappingEntry(3, 6, 25))
        insertLessonMappingEntry(LessonMappingEntry(3, 7, 26))
        insertLessonMappingEntry(LessonMappingEntry(3, 8, 27))
        insertStatisticsEntry(StatisticEntry(1, 1, 1, 0.0, Date()))
        insertStatisticsEntry(StatisticEntry(2, 1, 1, 0.0, Date()))
        insertStatisticsEntry(StatisticEntry(3, 1, 1, 0.0, Date()))
        insertStatisticsEntry(StatisticEntry(4, 1, 1, 0.0, Date()))
        insertStatisticsEntry(StatisticEntry(5, 1, 1, 0.0, Date(Date().time - 86400000)))
        insertStatisticsEntry(StatisticEntry(6, 1, 1, 0.0, Date(Date().time - 86400000)))
        insertStatisticsEntry(StatisticEntry(7, 1, 1, 0.0, Date(Date().time - 86400000 * 2)))
        insertStatisticsEntry(StatisticEntry(8, 1, 1, 0.0, Date(Date().time - 86400000 * 2)))
        insertStatisticsEntry(StatisticEntry(9, 1, 1, 0.0, Date(Date().time - 86400000 * 2)))
        insertStatisticsEntry(StatisticEntry(10, 1, 1, 0.0, Date(Date().time - 86400000 * 4)))
        insertStatisticsEntry(StatisticEntry(11, 1, 1, 0.0, Date(Date().time - 86400000 * 4)))
        insertStatisticsEntry(StatisticEntry(12, 1, 1, 0.0, Date(Date().time - 86400000 * 4)))
        insertStatisticsEntry(StatisticEntry(13, 1, 1, 0.0, Date(Date().time - 86400000 * 4)))
        insertStatisticsEntry(StatisticEntry(14, 1, 1, 0.0, Date(Date().time - 86400000 * 5)))
        insertStatisticsEntry(StatisticEntry(15, 1, 1, 0.0, Date(Date().time - 86400000 * 7)))
        insertStatisticsEntry(StatisticEntry(16, 1, 1, 0.0, Date(Date().time - 86400000 * 7)))
        insertStatisticsEntry(StatisticEntry(17, 1, 1, 0.0, Date(Date().time - 86400000 * 8)))
        insertStatisticsEntry(StatisticEntry(18, 1, 1, 0.0, Date(Date().time - 86400000 * 9)))
        insertStatisticsEntry(StatisticEntry(19, 1, 1, 0.0, Date(Date().time - 86400000 * 9)))
        insertStatisticsEntry(StatisticEntry(20, 1, 1, 0.0, Date(Date().time - 86400000 * 10)))
        insertStatisticsEntry(StatisticEntry(21, 1, 1, 0.0, Date(Date().time - 86400000 * 10)))
        insertStatisticsEntry(StatisticEntry(18, 1, 1, 0.0, Date(Date().time - 86400000 * 10)))
        insertStatisticsEntry(StatisticEntry(22, 1, 1, 0.0, Date(Date().time - 86400000 * 10)))
        //print the current entries in the db to the log console.
        mExecutors.diskIO().execute { Log.d(TAG, "Database has now " + mLibraryDao.getLibrarySize() + " library entries") }
        mExecutors.diskIO().execute {
            Log.d(TAG, "Database has now " + mLanguageDao.getAmountOfLanguages() + " language entries")
        }
        mExecutors.diskIO().execute { Log.d(TAG, "Database has now " + mLessonDao.getAmountOfLessons() + " lesson entries") }
        mExecutors.diskIO().execute { Log.d(TAG, "Database has now " + mLessonDao.getAmountOfLessonsMappings() + " lesson mapping entries") }
    }

    /**
     * Adds the given [LessonEntry]s to the Database.
     * If the [LessonEntry] already exists, nothing is done.
     *
     * @param callback      called with the ids of the added entries.
     * @param lessonEntries One or more [LessonEntry]s to add to the Database
     */
    fun insertLessonEntry(callback: OnLessonEntryAddedListener, vararg lessonEntries: LessonEntry) {
        mExecutors.diskIO().execute { callback.onLessonEntryAdded(mLessonDao.insertEntry(*lessonEntries)) }
    }

    /**
     * Get all [LibraryEntry]s.
     *
     * @return All [LibraryEntry]s.
     */
    fun getAllLibraryEntries(): LiveData<List<LibraryEntry>> = mLibraryDao.getAllEntries()

    /**
     * Get all [LibraryEntry]s.
     *
     * @return The [LibraryEntry] with the given ID.
     */
    fun getLibraryEntry(id: Int): LiveData<LibraryEntry> = mLibraryDao.getEntryById(id)

    /**
     * Adds the given [LibraryEntry]s to the Database.
     *
     * @param libraryEntry One or more [LibraryEntry]s to add to the Database
     */
    fun insertLibraryEntry(vararg libraryEntry: LibraryEntry) {
        mExecutors.diskIO().execute { mLibraryDao.insertEntry(*libraryEntry) }
        for (lib in libraryEntry) {
            Log.d(TAG, "Added Entry to db:$lib")
        }
    }

    /**
     * Removes the given [LibraryEntry]s from the Database.
     *
     * @param libraryEntry The @[LibraryEntry]s to remove from the Database
     */
    fun deleteLibraryEntry(libraryEntry: LibraryEntry) {
        mExecutors.diskIO().execute { mLibraryDao.deleteEntry(libraryEntry) }
    }

    /**
     * Updates the given [LibraryEntry]s to the Database.
     *
     * @param libraryEntries One or more [LibraryEntry]s to update in the Database
     */
    fun updateLibraryEntry(vararg libraryEntries: LibraryEntry) {
        mExecutors.diskIO().execute { mLibraryDao.updateEntry(*libraryEntries) }
    }

    /**
     * Get all [LanguageEntry]s.
     *
     * @return All [LanguageEntry]s.
     */
    fun getAllLanguageEntries(): LiveData<List<LanguageEntry>> = mLanguageDao.getAllEntries()

    /**
     * Removes the given [LanguageEntry]s from the Database.
     *
     * @param languageEntry The @[LanguageEntry]s to remove from the Database
     */
    fun deleteLanguageEntry(languageEntry: LanguageEntry) {
        mExecutors.diskIO().execute { mLanguageDao.deleteEntry(languageEntry) }
    }

    /**
     * Adds the given [LanguageEntry]s to the Database. If the language already exists,
     * nothing is done.
     *
     * @param languageEntries One or more [LanguageEntry]s to add to the Database
     */
    fun insertLanguageEntry(vararg languageEntries: LanguageEntry) {
        mExecutors.diskIO().execute { mLanguageDao.insertEntry(*languageEntries) }
    }

    /**
     * Get all [LessonEntry]s.
     *
     * @return All [LessonEntry]s.
     */
    fun getAllLessonEntries(): LiveData<List<LessonEntry>> = mLessonDao.getAllEntries()

    /**
     * Gets the active Language [LanguageEntry]s in the language database table.
     *
     * @return [LiveData] with the active @[LanguageEntry]s. Null if none is active.
     */
    fun getActiveLanguage(): LiveData<LanguageEntry?> = mLanguageDao.getActiveLanguage()

    /**
     * Sets the active language in the database.
     *
     * @param language The active language
     */
    fun setActiveLanguage(language: String) {
        mExecutors.diskIO().execute { mLanguageDao.setActiveLanguage(language) }
    }

    /**
     * Updates the given [LessonEntry]s to the Database.
     *
     * @param lessonEntries One or more [LessonEntry]s to update in the Database
     */
    fun updateLessonEntry(vararg lessonEntries: LessonEntry) {
        mExecutors.diskIO().execute { mLessonDao.updateEntry(*lessonEntries) }
    }

    /**
     * Removes the given [LessonEntry]s from the Database.
     *
     * @param lessonEntry The @[LessonEntry]s to remove from the Database
     */
    fun deleteLessonEntry(lessonEntry: LessonEntry) {
        mExecutors.diskIO().execute { mLessonDao.deleteEntry(lessonEntry) }
    }

    /**
     * Get the [LessonEntry] with the given ID.
     *
     * @return The [LessonEntry] with the given ID.
     */
    fun getLessonEntry(id: Int): LiveData<LessonEntry> = mLessonDao.getEntryById(id)

    /**
     * Adds the given [LessonMappingEntry]s to the Database. If the
     * [LessonMappingEntry] already exists, nothing is done.
     *
     * @param lessonMappingEntries One or more [LanguageEntry]s to add to the Database
     */
    fun insertLessonMappingEntry(vararg lessonMappingEntries: LessonMappingEntry) {
        mExecutors.diskIO().execute { mLessonDao.insertMappingEntry(*lessonMappingEntries) }
    }

    /**
     * Deletes a mapping between a [LessonEntry] and a [LibraryEntry].
     *
     * @param lessonMappingEntry The [LessonMappingEntry] describing the mapping.
     */
    fun deleteLessonMappingEntry(lessonMappingEntry: LessonMappingEntry) {
        mExecutors.diskIO().execute {
            mLessonDao.deleteMappingEntry(lessonMappingEntry.lessonEntryId, lessonMappingEntry.libraryEntryId)
        }
    }

    /**
     * Gets the List of [MappingPOJO]s in the lesson database table with the given id
     * based on the current language.
     *
     * @param id The lesson id for which the [List] of [MappingPOJO]s
     * should be retrieved.
     * @return [LiveData] with the [List] of @[MappingPOJO]s.
     */
    fun getMappingEntries(id: Int): LiveData<List<MappingPOJO>> = mLessonDao.getLessonMappingById(id)

    fun getTrainingEntries(number: Int, minKnowledgeLevel: Double, maxKnowledgeLevel: Double, lessonId: Int): LiveData<List<LibraryEntry>> =
            mLibraryDao.getTrainingEntries(number, minKnowledgeLevel, maxKnowledgeLevel, lessonId)

    fun getTrainingEntries(number: Int, minKnowledgeLevel: Double, maxKnowledgeLevel: Double): LiveData<List<LibraryEntry>> =
            mLibraryDao.getTrainingEntries(number, minKnowledgeLevel, maxKnowledgeLevel)

    /**
     * Adds the given [StatisticEntry]s to the Database.
     *
     * @param statisticEntries One or more [StatisticEntry]s to add to the Database
     */
    fun insertStatisticsEntry(vararg statisticEntries: StatisticEntry) {
        mExecutors.diskIO().execute { mStatisticsDao.insertEntry(*statisticEntries) }
    }

    /**
     * Gets a [List] of [StatisticsLibraryWordCount] entries, where each entry
     * contains the amount of words for each knowledgeLevel range (0-0.9999, 1-1.9999,...).
     *
     * @return The List of [StatisticsLibraryWordCount]s
     */
    fun getStatisticsLibraryCountByKnowledgeLevel(): LiveData<List<StatisticsLibraryWordCount>> =
            mStatisticsDao.getStatisticsLibraryCountByKnowledgeLevel()

    /**
     * Returns the activity during the last 14 days. For each active day a
     * [StatisticsActivityEntry] is added to the list. Days without activity have no entry.
     *
     * @return The List of [StatisticsActivityEntry] for all active days.
     */
    fun getStatisticsActivity(): LiveData<List<StatisticsActivityEntry>> {
        val calendar = Calendar.getInstance()
        calendar.time = Date()
        calendar.add(Calendar.DAY_OF_YEAR, -14)
        return mStatisticsDao.getStatisticsActivity(calendar.time)
    }

    /**
     * Get the date of the last training.
     *
     * @return Date of the last training wrapped in LiveData.
     */
    fun getLastTrainingDate(): LiveData<StatisticsLastTrainingDate> = mStatisticsDao.getLastTrainingDate()

    /**
     * Returns the lesson with the lowest KnowledgeLevel
     *
     * @return LessonKnowledgeLevel holding the result wrapped in LiveData.
     */
    fun getWorstLesson(): LiveData<LessonKnowledgeLevel> = mStatisticsDao.getWorstLesson()
}